File: src/fromKrisz.c
// Author: Krisztian Sebestyen ksebestyen@gmail.com

//#include "matrix.h"
#include <float.h> //DBL_EPSILON
#include <R_ext/Lapack.h>
#include <R.h>
//#include <R_ext/Applic.h>
//#include <R_ext/BLAS.h>
//#include <R_ext/RS.h> //R definitions for 'extending' R, registering functions,...

#define PRINTF Rprintf
#define	MAX(A,B)	((A) > (B) ? (A) : (B))
#define	MIN(A,B)	((A) < (B) ? (A) : (B))


// LEADING DIMENSION 'LDA' or 'LD', comments collected from the 'net'
// 'ld' : 'leading dimension of matrix' is equal to the number of elements in its major dimension.
//      : the distance between two neighboring elements along minor dimension.
// It is used to access a sub-matrix of a larger matrix
// The leading dimension for a two-dimensional array is an increment that is used to find the 
// starting point for the matrix elements in each successive column of the array. 
// The leading dimension must always be positive. It must always be greater than or equal to the
// minor dimension of the matrix.
#define CX(i,j,ld) (((j)*(ld))+(i)) //indexing matrices in column-major order
#define RX(i,j,ld) (((i)*(ld))+(j)) //indexing matrices in row-major order
typedef enum {COLUMN_MAJOR = 0,ROW_MAJOR = 1}MAJOR_ORDER;



// matrix multiplication D %*% X %*% D2 for D,D2 diagonal matrices and X square matrix in column-major order
// x,y are square matrices in column order, d1 and d2 represent the diagonals of diagonal matrices
void dxd_(int* _n, double* d1,double* x,double* d2,double* result){
	int i,j;
	int n = *_n;
	for(j = 0;j < n;j++)
		for(i = 0;i < n;i++)
			result[j*n + i] = d1[i] * x[j*n + i] * d2[j];
}

void dxd(int* n, double* d1,double* _x,double* d2,double* _y){
	int i,j;
	double* x = _x;
	double* y = _y;
	for(j = 0;j < *n;j++){
		for(i = 0;i < *n;i++){
			*y = d1[i] * (*x) * d2[j];
			y++;x++;
		}
	}
}


void symprod(int* nrow,int* ncol,double* S,double* X,double* Y)
{

	char*  	SIDE = "L";
	char*  	UPLO = "U";
	double alpha = 1.0;
	double beta = 0.0;
	
	F77_CALL(dsymm)( 	
		SIDE,UPLO,
		nrow,ncol,
		&alpha,
		S,nrow,
		X,nrow,
		&beta,
		Y,nrow 
	); 
	
}


void txSy(int* n,double* S,double* x,double* y,double* temp,double* out){

	double alpha = 1.0;
	double beta = 0.0;
	int ione = 1;
	F77_CALL(dsymv)("U",n,&alpha,S,n,y,&ione,&beta,temp,&ione); // temp = S %*% y
	*out = F77_CALL(ddot)(n, x, &ione, temp, &ione); // out = x %*% temp
}

File: R/fromKrisz.R
## Author: Krisz Sebestyen
# matrix functions from Krisz used to improve performance of krm.score.test

# .as.double <- as.double
# as.double(x) for older versions of R when using .C(DUP = FALSE) make duplicate copy of x
# in addition, even if x is a 'double', since x has attributes (dim(x)) as.double(x) duplicates
.as.double <- function(x, stripAttributes=FALSE){
    if(!stripAttributes){
        if(is.double(x)) return(x)   # no duplicate copy
        storage.mode(x) <- 'double'  # yes duplicate copy
        return(x)
    }
    return(as.double(x))             # duplicate copy if has attributes or not double 
}

DXD <- function(d1,x,d2){ 
    
    if(NROW(d1) == NCOL(d1)) d1 <- diag(d1) # reduced to diagonal of the matrix
    if(NROW(d2) == NCOL(d2)) d2 <- diag(d2) # reduced to diagonal of the matrix
    
    n<-length(d1)
    
    if((NROW(x) != n) || (NCOL(x) != n)) 
    if(!is.matrix(x)) x <- as.matrix(x)
    
    
    d1 <- .as.double(d1) # critical to use .as.double here, otherwise it will try to copy
    d2 <- .as.double(d2)
    x  <- .as.double(x)
    
    .C("dxd",n,d1,x,d2,dxd = x*0.0,DUP = FALSE,NAOK = FALSE)$dxd
}


# Y = S %*% X, S symmetric (m x m), X,Y (m x n)
symprod <- function(S,X, Y = NULL){

    if(!is.matrix(S) || !is.matrix(X)) stop("Both 'S' and 'X' have to be matrices.")
    S <- .as.double(S)
    X <- .as.double(X)
    if(is.null(Y)) Y <- X * 0.0
    if(
        (NCOL(S) != NROW(X)) || 
        (NCOL(S) != NROW(S)) || 
        !all(dim(X) == dim(Y))
    ) stop("Dimension mismatch")
    
    res <- .C( 
        "symprod",  
        M = NROW(Y),
        N = NCOL(Y),
        A = .as.double(S),
        B = as.double(X),
        C = Y
    ) 

    # res <- .Fortran( 
        # "dsymm",  
        # SIDE = 'L',
        # UPLO = 'U',
        # M = NROW(Y),
        # N = NCOL(Y),
        # alpha = as.double(1),
        # A = .as.double(S),
        # LDA = NROW(S),
        # B = as.double(X),
        # LDB = NROW(X),
        # beta = as.double(0),
        # C = Y,
        # LDC = NROW(Y) 
    # )     
    res$C
}


txSy <- function(x,S,y){
    n <- length(x)
    
    if( (n != NROW(S)) || (n != NCOL(S)) || (n != length(y)) )stop("Dimension mismatch")
    
    .C('txSy',length(x),.as.double(S),.as.double(x),.as.double(y),double(n),out=double(1))$out
}
